core/ir/encode_shared.c

```diff
@@ -139,5 +140,14 @@ byte *
 instr_encode(void *drcontext, instr_t *instr, byte *pc)
 {
     dcontext_t *dcontext = (dcontext_t *)drcontext;
+
+    /* Synthetic ISA has its own encoder.
+     * XXX i#1684: when DR can be built with full dynamic architecture selection we won't
+     * need to pollute the encoding of other architectures with this synthetic ISA special
+     * case.
+     */
+    if (instr_get_isa_mode(instr) == DR_ISA_SYNTHETIC)
+        return encode_to_synth(dcontext, instr, pc);
```

Comment:
Check for instr_t synthetic ISA moved to instr_encode() from instr_encode_arch().
This avoid a potential loop where encode_to_synth() can potentially call itself: encode_to_synth()->instr_get_arith_flags()->instr_get_eflags()->private_instr_encode()->instr_encode_check_reachability()->instr_encode_arch()->encode_to_synth().

Reply:
Moved encode_to_synth() from instr_encode() to instr_encode_to_copy(). https://github.com/DynamoRIO/dynamorio/pull/6691#discussion_r1533142127